### ISSUE TITLE
Error in Ghidra plugin, after importing the classes

### ISSUE THREAD

**Issue by 0xBEEEF** [cmu-sei/pharos#125]
I have successfully analyzed a very large project with about 9000 classes with the OOAnalyzer. Also I was able to create the JSON file for the import to Ghidra without any problems and import at least the first part of it.

The first steps work without problems, but when the classes are imported and then assigned in the second step, the first thing you get is the success message that the 9000 classes have been imported.

In the background there are still some tasks running which seem to finish something. In any case, I can't save the project, because I always get the following message (can sometimes be more than one task, as shown in this example)

![image](https://user-images.githubusercontent.com/36008527/88074634-f3cbfc00-cb77-11ea-9655-1d04a64a2ec9.png)

And then suddenly, out of nowhere, the following hard error appears.

> An unexpected error occurred while processing the command: OOAnalyzer Import
> java.lang.IllegalArgumentException: Data type TestClassT can't contain itself.
> 	at ghidra.program.database.data.CompositeDB.checkAncestry(CompositeDB.java:265)
> 	at ghidra.program.database.data.StructureDB.doReplaceWithUnaligned(StructureDB.java:1194)
> 	at ghidra.program.database.data.StructureDB.doReplaceWith(StructureDB.java:1133)
> 	at ghidra.program.database.data.DataTypeManagerDB.createStructure(DataTypeManagerDB.java:2160)
> 	at ghidra.program.database.data.DataTypeManagerDB.createDataType(DataTypeManagerDB.java:2099)
> 	at ghidra.program.database.data.DataTypeManagerDB.resolveNoSourceDataType(DataTypeManagerDB.java:847)
> 	at ghidra.program.database.data.DataTypeManagerDB.resolve(DataTypeManagerDB.java:735)
> 	at ghidra.program.database.data.DataTypeManagerDB.addDataType(DataTypeManagerDB.java:1005)
> 	at ooanalyzer.OOAnalyzer.updateTypeManager(OOAnalyzer.java:1419)
> 	at ooanalyzer.OOAnalyzer.analyzeClasses(OOAnalyzer.java:434)
> 	at ooanalyzer.OOAnalyzerGhidraPlugin$ImportCommand.cmdConfigureAndExecute(OOAnalyzerGhidraPlugin.java:166)
> 	at ooanalyzer.OOAnalyzerGhidraPlugin$ImportCommand.applyTo(OOAnalyzerGhidraPlugin.java:99)
> 	at ghidra.framework.plugintool.mgr.BackgroundCommandTask.run(BackgroundCommandTask.java:101)
> 	at ghidra.framework.plugintool.mgr.ToolTaskManager.run(ToolTaskManager.java:315)
> 	at java.base/java.lang.Thread.run(Thread.java:835)
> 
> ---------------------------------------------------
> Build Date: 2020-Feb-12 1149 EST
> Ghidra Version: 9.1.2
> Java Home: C:\Tools\openjdk-12.0.2_windows-x64_bin\jdk-12.0.2
> JVM Version: Oracle Corporation 12.0.2
> OS: Windows 10 10.0 amd64
> Workstation: Demo

But the tasks are still executed in the background, and everything imported, at least the function assignments or class assignments of the second step are removed. Seems to be some kind of rollback, and only affects the second step, but according to the message "xyz classes have been successfully imported" it should already be completed.

But I can neither save the project, see the picture above, nor somehow find out which tasks are responsible for it in the background.

To my regret I can't provide the original file or the generated JSON file here. Nevertheless I hope that the bug can be fixed somehow.

For the import to Ghidra I used the following version of the plugin:
ghidra_9.1.2_PUBLIC_20200718_OOAnalyzerPlugin

**Comment by sei-eschwartz** [comment]
We should be able to handle the exception which should allow you to save the database.  I'll work on that shortly.

I am not sure why there would be a recursive reference like this though.  If you are able to share a snippet of your .results or .json file that deals with `TestClassT`, and specifically how it includes a reference to itself, that would be very helpful.  If not, no worries.  We understand that some targets are sensitive.

**Comment by sei-eschwartz** [comment]
Can you try this? https://github.com/sei-eschwartz/pharos/tree/issue125

Built plugin is available here: https://dev.azure.com/seipharos/pharos/_build/results?buildId=201&view=artifacts&type=publishedArtifacts

**Comment by 0xBEEEF** [comment]
Thank you for checking the bug so quickly, and for providing a new version so quickly. Unfortunately, today I can't get around to checking if this solves the mentioned bug. I will probably not be able to do this until tomorrow or the day after tomorrow.

Personally, I think it's gonna be difficult.

What I can think of that might help you more. You should be able to catch this exception somehow. Could you perhaps provide some information about the environment or objects at that time, which I can then give you? The context is the most important thing in this case.

The only thing I noticed while checking this file is that there are many similar classes with the same name, but they live in different namespaces. Could it be that this leads to the problem at some point? Otherwise I could not detect such a recursion within the JSON file. Or can you give me some hints about the file I should check for these classes?

**Comment by sei-eschwartz** [comment]
> Thank you for checking the bug so quickly, and for providing a new version so quickly. Unfortunately, today I can't get around to checking if this solves the mentioned bug. I will probably not be able to do this until tomorrow or the day after tomorrow.

That's fine.

> What I can think of that might help you more. You should be able to catch this exception somehow. Could you perhaps provide some information about the environment or objects at that time, which I can then give you? The context is the most important thing in this case.

I did add a warning message here: https://github.com/sei-eschwartz/pharos/commit/20222eca1fe4846fefd97240f0f8db0dcf058b90#diff-d777d1dba8af8f3d42f60251ff516031R1397

If you could report that back it might be helpful.

> The only thing I noticed while checking this file is that there are many similar classes with the same name, but they live in different namespaces. Could it be that this leads to the problem at some point? Otherwise I could not detect such a recursion within the JSON file. Or can you give me some hints about the file I should check for these classes?

Interesting.  That could be the problem.  We did have some code that ignored namespaces if it couldn't find the fully qualified name.  I forget if I removed it.

**Comment by 0xBEEEF** [comment]
Good news! I've actually gotten around to testing the fix you provided yesterday, and as a little treat, I also got the anonymized information from the JSON file that you wanted.

But let's start at the beginning. I was able to install the plugin, and tried it in a completely new version of the last database (rebuilt for safety).

The good news is that the mentioned bug is actually caught. The bad news is that the message you marked in the commit is not shown to me at all. The program runs smoothly through all the phases, but in the end nothing happened.

In yesterday's bug-prone version, at least after the bug, the structures were still present in the database. But I can't even find them. Not to mention the fact that at the end of the process everything looks as if nothing at all has happened. When I ran the process in the buggy version, I could stand in a function live and see how things change in the system. Now it flickers from time to time, but nothing really happens anymore. Do you have any idea why nothing seems to be happening now?

To the [JSON file](https://pastebin.com/Z92eTumt). I have extracted the essence of it across all data types that have any relation to the mentioned class. As you can already see, there are many different namespaces where the same class lives. Some of them are also nested. But this is only my abstraction. I did not spend hours on the JSON file. I have reduced the data to the most important, let's hope that nothing important is missing that you need for your analysis.

**Comment by sei-eschwartz** [comment]
> The good news is that the mentioned bug is actually caught. The bad news is that the message you marked in the commit is not shown to me at all. The program runs smoothly through all the phases, but in the end nothing happened.

This should appear in `application.log`.  On my linux system, this is located at `~/.ghidra/.ghidra_9.1.2_PUBLIC/application.log`.

> In yesterday's bug-prone version, at least after the bug, the structures were still present in the database. But I can't even find them. Not to mention the fact that at the end of the process everything looks as if nothing at all has happened. When I ran the process in the buggy version, I could stand in a function live and see how things change in the system. Now it flickers from time to time, but nothing really happens anymore. Do you have any idea why nothing seems to be happening now?

Are you sure it finished importing?  It can take a long time.  It will pop up a dialog box when it's done.  Generally the flashing is a sign that it's still doing something.

> To the [JSON file](https://pastebin.com/Z92eTumt). I have extracted the essence of it across all data types that have any relation to the mentioned class. As you can already see, there are many different namespaces where the same class lives. Some of them are also nested. But this is only my abstraction. I did not spend hours on the JSON file. I have reduced the data to the most important, let's hope that nothing important is missing that you need for your analysis.

Thank you.  I think I understand the root cause, and should be able to make my own synthetic test case that triggers the problem.

**Comment by 0xBEEEF** [comment]
Yes, the tasks have all been completed. The success message also came at the end as you described it. This time I was able to save the result even without an error concerning a still running task. But obviously nothing happened on the surface. But unfortunately I couldn't find anything even with the search, not even in the same functions that were filled last time at least during the process. Everything somehow strange. It's very mysterious. Yesterday there was even in the structures of the program some folder named "OOAnalyzer". Not even that was there. Because of the log, I'll see if it's more revealing tomorrow.

**Comment by sei-eschwartz** [comment]
That is very odd.  If the import completed, you should absolutely have a "OOAnalyzer" folder in both the symbol tree and the data type manager.  Looking at `application.log` is probably the best thing to do.  It's fairly verbose and will hopefully give us a clue.

**Comment by 0xBEEEF** [comment]
So here I am again. I was able to find the log you mentioned - under Windows it is located somewhere else, but I found it.

Yesterday I did the import twice in a row, because I didn't notice any changes as described above. In between I was able to save the database without a hint that there are still running tasks. That means you seem to catch the exceptions correctly after all.

The log gives a good information about this. After the import process starts again it creates exactly the same structures, classes and namespaces. All the structures already created in the first import do not seem to exist. What I should also mention. My file uses Ghidra's versioning feature to allow you to make changes and possibly undo them.

What else did I notice now? Today I imported the file again and put it back into version control after Ghidra's own analysis. During the import process only the icon window is flickering. And here I can actually see that there is a new namespace "OOAnalyzer". This namespace is always being extended. Since the process is still running and this will take some hours I can't say anything more about it. In any case, something is obviously happening.

In the Datatype Manager window I could not find any changes, apparently nothing new has been added yet.

I [uploaded](https://pastebin.com/ekqudTX4) the log anonymously - at least the most important parts of it, where you can see the two successive analyses very well. Therefore important numbers are missing. Everywhere where 0 is, there used to be a number - so don't be worried that something is not filled here. In between the analyses (always at the end) Ghidra does another auto analysis. I can't say if this is necessary/should be done that way.

I hope the log helps you a little bit. 

Oh, if I should try something else. Unfortunately it will not work until next week, I won't have the chance to do it before then.

**Comment by sei-eschwartz** [comment]
It sounds like Ghidra's versioning feature might be interfering.  I haven't used that before so I'll experiment with that and see if I can reproduce the issue.

**Comment by sei-eschwartz** [comment]
I tried the following using version control:

Version 1: Initial import of file
Version 2: Ghidra auto analysis
Version 3: OOAnalyzer import

I was able to go back to version 2 and reimport without issue.

So I'm at a loss as to why you are having trouble.  It sounds like reimporting the file from scratch is working though, which is good to hear.

**Comment by 0xBEEEF** [comment]
All right, thanks for cross-checking.  Personally, I'd much rather that this only happens to me and the error does not exist everywhere. My suggestion: I'll just take a complete look at it again next week. For this I'll use a completely new environment, an empty operating system. But Windows has to stay that way. Here I install everything completely new and test it again in peace. 

Regarding your linked changes that require an adjustment to the JSON file: To what extent do I have to carry out the process completely again? Is it enough to simply recreate the JSON (last step), or do I have to do the more extensive step of object analysis (penultimate step), or is it even necessary to start from scratch?

**Comment by sei-eschwartz** [comment]
> All right, thanks for cross-checking. Personally, I'd much rather that this only happens to me and the error does not exist everywhere. My suggestion: I'll just take a complete look at it again next week. For this I'll use a completely new environment, an empty operating system. But Windows has to stay that way. Here I install everything completely new and test it again in peace.

That sounds reasonable.  If you can identify the circumstances that lead to the problem happening, hopefully we'll have a better idea what the problem is and how to fix it.

> Regarding your linked changes that require an adjustment to the JSON file: To what extent do I have to carry out the process completely again? Is it enough to simply recreate the JSON (last step), or do I have to do the more extensive step of object analysis (penultimate step), or is it even necessary to start from scratch?

You can just rerun oojson.  There is no need to redo the more time consuming steps.

**Comment by 0xBEEEF** [comment]
So, I took another look at everything today. I started with a very fresh Windows installation. There I installed the JDK, and then Ghidra. 

Up to here everything went smoothly. Now I imported the program again and did the first analysis. After a few hours it was finished and I could start with the results from Pharos. These I imported into the database. So far so good, everything was imported one after the other. You could also see in the listing that changes were taking place. These were marked in the graphic as dark grey boxes. Yes, even the data types were imported, and I could already see the new data types in the Data Type Manager.

Well, but then came the very last commit of the outermost transaction (as I know by now) and poof, everything was back to the original state. I did another run afterwards, but this time without the "Namespace" option. Same behavior here. In the end, everything was gone. In the meantime I have found the place in the code of the Ghidra plugin where the behaviour occurs erroneously. Here you can find the marker.

https://github.com/cmu-sei/pharos/blob/7f168f7526f74622dce7eef9db121a0fe09456b5/tools/ooanalyzer/ghidra/OOAnalyzerPlugin/src/main/java/ooanalyzer/OOAnalyzerGhidraPlugin.java#L181

If this code is run through and Result > 0, which is true in this case, the transaction should be committed. Well, unfortunately not here, this is where the rollback takes place. Why I do not know yet, only the position is clear to me.

But before that, the success message is displayed in this line.

https://github.com/cmu-sei/pharos/blob/7f168f7526f74622dce7eef9db121a0fe09456b5/tools/ooanalyzer/ghidra/OOAnalyzerPlugin/src/main/java/ooanalyzer/OOAnalyzerGhidraPlugin.java#L174

**Comment by sei-eschwartz** [comment]
Very odd.  I will try to make a branch with some additional debugging output for the transactions.

I assume you don't see any evidence of exceptions, and that you are running Ghidra 9.1.2?

**Comment by 0xBEEEF** [comment]
No, the only exceptions I got were in the logfile I provided a few days ago. Oh, what else I noticed, I don't know, maybe this is important.

In the end, just before the system is completely finished, several tasks are running in parallel. I noticed this when I tried to save during the last phase of the analysis.

My guess is, since there are also some things going on in parallel, that there are still running tasks in the background that are not finished properly. But this is only a guess, I have no proof for it. That might explain some things better. In the code I can't see this without further ado, but I can't see where things are running in parallel here. Because the message is only shown to me very briefly on the screen. And directly after that the Auto Analysis is running again (in a shortened form, but it is running). Of course it has nothing to do, because nothing happened.

**Comment by sei-eschwartz** [comment]
What message is shown on the screen?

Did you try to save before the success message displayed?

**Comment by sei-eschwartz** [comment]
Please try this: https://dev.azure.com/seipharos/pharos/_build/results?buildId=204&view=artifacts&type=publishedArtifacts
https://github.com/cmu-sei/pharos/compare/master...sei-eschwartz:transaction_debugging

**Comment by 0xBEEEF** [comment]
Unfortunately I forgot to take a picture concerning the tasks. In any case I will make up for that. If I would have tried to save, then the rollback would have happened, because all running tasks would have been aborted. That was even written in the message. The message was the same as in the original issue message, but with more tasks.

On the topic "automatic analysis" I found the following closed [issue](https://github.com/cmu-sei/pharos/issues/79#issuecomment-633693094) in the repository. @astrelsky has written something about it here. Maybe he also has an idea what leads to this behaviour with the uncommitted data. In connection with Ghidra he seems to have some knowledge already.

By the way, in the same issue someone already reported the problem with the type that contains itself. Apparently there was a different behaviour between Windows and Linux. I'll check this out later if I have less problems under Linux when using this plugin.

About the exception: Unfortunately, it was only in the log. It was not displayed on the screen. Apparently the process couldn't set something because the stack didn't fit.

Oh, and about the environment. Yes, I use version 9.1.2 from February of this year.

**Comment by 0xBEEEF** [comment]
> Please try this: https://dev.azure.com/seipharos/pharos/_build/results?buildId=204&view=artifacts&type=publishedArtifacts
> [master...sei-eschwartz:transaction_debugging](https://github.com/cmu-sei/pharos/compare/master...sei-eschwartz:transaction_debugging)

Thanks, I'll try it out as soon as I can.

**Comment by astrelsky** [comment]
https://github.com/cmu-sei/pharos/blob/7f168f7526f74622dce7eef9db121a0fe09456b5/tools/ooanalyzer/ghidra/OOAnalyzerPlugin/src/main/java/ooanalyzer/OOAnalyzer.java#L1404-L1413

You should have a finally block which will end the transaction committing the changes only if no exception occurred. The transaction is never ended due to the exception being thrown in CompositeDB.

**Comment by sei-eschwartz** [comment]
> Unfortunately I forgot to take a picture concerning the tasks. In any case I will make up for that. If I would have tried to save, then the rollback would have happened, because all running tasks would have been aborted. That was even written in the message. The message was the same as in the original issue message, but with more tasks.

I would be interested if you were to test without trying to save during the import.  Maybe there is a Ghidra bug that interfered with the transactions somehow even if you did not force the save.

> On the topic "automatic analysis" I found the following closed [issue](https://github.com/cmu-sei/pharos/issues/79#issuecomment-633693094) in the repository. @astrelsky has written something about it here. Maybe he also has an idea what leads to this behaviour with the uncommitted data. In connection with Ghidra he seems to have some knowledge already.

I don't see any reason why automatic analysis should not run after the OOAnalyzer import.  I don't consider this a bug.

> By the way, in the same issue someone already reported the problem with the type that contains itself. Apparently there was a different behaviour between Windows and Linux. I'll check this out later if I have less problems under Linux when using this plugin.

If I recall correctly, the problem was user error and not a difference between Windows and Linux.

> About the exception: Unfortunately, it was only in the log. It was not displayed on the screen. Apparently the process couldn't set something because the stack didn't fit.

The java call stack or the stack of the executable being analyzed?  It would be helpful to know what this exception was.

> Oh, and about the environment. Yes, I use version 9.1.2 from February of this year.

Good, just wanted to check.

**Comment by sei-eschwartz** [comment]
> Edit: I didn't realize the IllegalArgumentException is being caught so I will look further. Regardless if any other exception is thrown you will encounter an issue.

Yes, perhaps what is going on is we try to commit the main transaction and one of the sub-transactions is still open.  I think the debug code I added should allow us to detect if that is happening.

**Comment by astrelsky** [comment]
> > Edit: I didn't realize the IllegalArgumentException is being caught so I will look further. Regardless if any other exception is thrown you will encounter an issue.
> 
> Yes, perhaps what is going on is we try to commit the main transaction and one of the sub-transactions is still open. I think the debug code I added should allow us to detect if that is happening.

In general you don't need to handle transactions. If you are using an Analyzer or BackgroundCmd the transactions are handled by the abstract class or the AutoAnalysisManager. It is usually when dealing with a file datatype manager that it becomes necessary to handle them.

**Comment by 0xBEEEF** [comment]
There's news! I have now imported the Ghidra extension with the latest version of the JSON file.

It turned out that the assumption with the unfinished transactions was correct. You can see it quite well in the log. I uploaded the log [here](https://pastebin.com/N5N6FYjL), at least the most important parts.

Maybe this will help in the search. What else I noticed, and this is also visible in the log, is that there are some exceptions that apparently were not caught. These were not shown to me in the GUI, but were only in the log. So I assume that @astrelsky is right in saying that something is wrong in the process, which finally leads to uncommitted data.

And something else new. I tried to recreate the complete process on a Linux system - my main concern was to make sure that the behaviour is identical. Luckily this was the case. So we can definitely rule out that it's something operating system specific. I would also have been very surprised.

**Comment by sei-eschwartz** [comment]
> There's news! I have now imported the Ghidra extension with the latest version of the JSON file.

Thank you.
 
> It turned out that the assumption with the unfinished transactions was correct. You can see it quite well in the log. I uploaded the log [here](https://pastebin.com/N5N6FYjL), at least the most important parts.

Those are the two unfinished transactions that are supposed to be there.  But the `Status: 3` is interesting, because it corresponds to `NOT_DONE_BUT_ABORTED`.  I will have to look at the Ghidra source code to see what could cause that (unless @astrelsky has any ideas).
 
> Maybe this will help in the search. What else I noticed, and this is also visible in the log, is that there are some exceptions that apparently were not caught. These were not shown to me in the GUI, but were only in the log. So I assume that @astrelsky is right in saying that something is wrong in the process, which finally leads to uncommitted data.

Actually those all look like they are caught.  The original exceptions are just printed by the exception handlers.

> And something else new. I tried to recreate the complete process on a Linux system - my main concern was to make sure that the behaviour is identical. Luckily this was the case. So we can definitely rule out that it's something operating system specific. I would also have been very surprised.

The more we can rule out the better!

I'm feeling optimistic that the `NOT_DONE_BUT_ABORTED` status might be a good lead.  I will write back after I do some research on that.

**Comment by astrelsky** [comment]
> Those are the two unfinished transactions that are supposed to be there. But the `Status: 3` is interesting, because it corresponds to `NOT_DONE_BUT_ABORTED`. I will have to look at the Ghidra source code to see what could cause that (unless @astrelsky has any ideas).
> 

All the exceptions do appear to be handled so I don't think that is the cause of the unfinished transactions. As a sanity check you could change cmdConfigureAndExecute to return a boolean and remove https://github.com/cmu-sei/pharos/blob/305f14786915ddc8ba4e235aa14a129b9ccd6508/tools/ooanalyzer/ghidra/OOAnalyzerPlugin/src/main/java/ooanalyzer/OOAnalyzerGhidraPlugin.java#L161 and the corresponding endTransaction. You are able to do this as long as the BackgroundCommand is executed properly via PluginTool.executeBackgroundCommand, which appears to be the case. PluginTool.executeBackgroundCommand will start and end the transaction for the domain object before/after invoking BackgroundCommand.applyTo and uses that result for the commit parameter during the endTransaction. This would merly help eliminate the problem resulting from something the plugin is doing.

**Comment by 0xBEEEF** [comment]
> All the exceptions do appear to be handled so I don't think that is the cause of the unfinished transactions.

All right!

> Actually those all look like they are caught. The original exceptions are just printed by the exception handlers.
> 
> >

So much the better! I was just trying to add any information that might have confused me in the log.

What just confuses me is that these transaction problems only occur in my example. As @sei-eschwartz  wrote above, he could not reproduce the behavior with his examples.

If it is really due to the misplacement or misuse of a transaction, then it should happen there as well. I personally do not understand that. Is there any other aspect that could lead to this behavior?

**Comment by sei-eschwartz** [comment]
I haven't made a lot of progress.  I did verify that when I run the branch, I get status 0, not 3.  So something is aborting the transaction.  My hunch is that something is calling `abort`:

https://sourcegraph.com/github.com/NationalSecurityAgency/ghidra@31a0857395751c61317e33dd8d52487634c12954/-/blob/Ghidra/Framework/Project/src/main/java/ghidra/framework/data/DomainObjectDBTransaction.java#L259

If I can figure out how, I'll try to add a listener to print something in the log exactly when the transaction is aborted.

**Comment by sei-eschwartz** [comment]
I wasn't able to figure out how to add a listener.  That class is private; I'm not sure it's possible.

My advice would be to compile Ghidra from source and add a debug message to the abort method (and anywhere else the status of the transaction is set to 2).

I am otherwise out of ideas :-(

**Comment by astrelsky** [comment]
> I wasn't able to figure out how to add a listener. That class is private; I'm not sure it's possible.
> 
> My advice would be to compile Ghidra from source and add a debug message to the abort method (and anywhere else the status of the transaction is set to 2).
> 
> I am otherwise out of ideas :-(

[DomainObjectAdapterDB.html.addTransactionListener(TransactionListener)](https://ghidra.re/ghidra_docs/api/ghidra/framework/data/DomainObjectAdapterDB.html#addTransactionListener(ghidra.framework.model.TransactionListener))
[DomainObjectAdapterDB.startTransaction(String, AbortedTransactionListener)](https://ghidra.re/ghidra_docs/api/ghidra/framework/data/DomainObjectAdapterDB.html#startTransaction(java.lang.String,ghidra.framework.model.AbortedTransactionListener))

You can get away with casting a Program to ProgramDB which is a DomainObjectAdapterDB.

**Comment by sei-eschwartz** [comment]
I don't think that listener class receives aborts.  😐

Sent from Workspace ONE Boxer

On Jul 29, 2020 6:21 PM, Andrew Strelsky <notifications@github.com> wrote:

I wasn't able to figure out how to add a listener. That class is private; I'm not sure it's possible.

My advice would be to compile Ghidra from source and add a debug message to the abort method (and anywhere else the status of the transaction is set to 2).

I am otherwise out of ideas :-(

DomainObjectAdapterDB.html.addTransactionListener(TransactionListener)<https://ghidra.re/ghidra_docs/api/ghidra/framework/data/DomainObjectAdapterDB.html#addTransactionListener(ghidra.framework.model.TransactionListener)>
DomainObjectAdapterDB.startTransaction(String, AbortedTransactionListener)<https://ghidra.re/ghidra_docs/api/ghidra/framework/data/DomainObjectAdapterDB.html#startTransaction(java.lang.String,ghidra.framework.model.AbortedTransactionListener)>

You can get away with casting a Program to ProgramDB which is a DomainObjectAdapterDB.

**Comment by astrelsky** [comment]
The AbortedTransactionListener does. If you wish to add one you can do so via reflection for debugging purposes. It would look similar to the following:

```java

private void addAbortListened(Program program) throws Exception {
    ProgramDB db = (ProgramDB) program;
    DomainObjectDBTransaction transaction = (DomainObjectDBTransaction) db.getCurrentTransaction();
    Field abortListenersField = transaction.getClass().getDeclaredField("abortedTransactionListeners");
    abortListenersField.setAccessible(true);
    abortListenersField.get(transaction).add(yourListener);
}
```

Just be advised that the abortedTransactionListeners field is a CopyOnWriteWeakSet. I don't think it matters though because you are reflectively accessing the transactions field. Obviously this isn't a great idea to leave in the plugin but can be helpful for debugging this.

**Comment by edmcman** [comment]
Thanks @astrelsky.  I'll give this a try.  I attempted to do something yesterday using reflection, but I don't think it worked.

**Comment by sei-eschwartz** [comment]
So that didn't quite work because `DomainObjectDBTransaction` can't be referenced outside of its package.

I almost have this working here:

https://github.com/sei-eschwartz/pharos/commit/ad19c66c39c6c2627127f65c353203e88a59c899

But I'm getting the following exception:

```
2020-07-30 10:30:58 WARN  (OOAnalyzerGhidraPlugin$ImportCommand) Exception java.lang.NoSuchMethodException: ghidra.util.datastruct.CopyOnWriteWeakSet.add(ghidra.framework.model.AbortedTransactionListener)
```

Seems like it might be related to the data structure being a generic?

**Comment by sei-eschwartz** [comment]
Nevermind, managed to fix that.  For generics the type is just `Object.class`.  Now getting:

```
2020-07-30 10:44:29 WARN  (OOAnalyzerGhidraPlugin$ImportCommand) Exception java.lang.IllegalAccessException: class ooanalyzer.OOAnalyzerGhidraPlugin$ImportCommand cannot 
access a member of class ghidra.util.datastruct.CopyOnWriteWeakSet with modifiers "public synchronized"
```

**Comment by sei-eschwartz** [comment]
Finally got it to work.  That error message was just a very bad way of saying the method is not accessible.  The "public synchronized" is a red herring.

I'll post a link to the built extension as soon as it's done.  @0xBEEEF  if you could try that and report when the transaction is aborted, that will hopefully get us some more information.

**Comment by sei-eschwartz** [comment]
https://dev.azure.com/seipharos/pharos/_build/results?buildId=209&view=artifacts&type=publishedArtifacts

Also thanks @astrelsky for all the help

**Comment by 0xBEEEF** [comment]
Thank you both for your efforts to get this problem under control. As usual, I performed analysis again and then searched for "listener" in the log, because this should mark the aborts. This info was taken from the commit.

But what did I find?  The text was only available twice, in one and the same line. And that at the very end. [Here](https://pastebin.com/V1qmfVh2) is the line with the nearer surroundings from the log where the text could be found.

Oh, I also checked the exceptions that occurred. Here also no new one was added. Everything here is unchanged.

Any more ideas how to get the problem under control?

**Comment by sei-eschwartz** [comment]
So my theory about the `abort` method being called was wrong.  I think I understand what is happening now.

I believe that the following code effectively aborts the parent transaction if one of the sub-transactions requests a roll-back:

https://sourcegraph.com/github.com/NationalSecurityAgency/ghidra@a77dcf72764ae0746b7e2e95fc1c8af67976033c/-/blob/Ghidra/Framework/Project/src/main/java/ghidra/framework/data/DomainObjectDBTransaction.java?utm_source=share#L157

As a test, I created a dummy transaction and aborted it.  Guess what happened?  I got the message about classes being imported, but they were not!  Sound familiar?

The original author of the plugin used transactions quite often, and it is clear that he did not expect this behavior.

In the short term, I will look at your logs of the exceptions and try to remove the transactions that are failing.

In the longer term, we need to remove the sub-transactions.  There really isn't much of a point of using them if we can't roll back without invalidating the parent transaction.

**Comment by sei-eschwartz** [comment]
The pastebin logs expired.  But fortunately I think there are only two sub-transactions and they are easy to get rid of.  They are both related to adding types to the data type manager, and I do recall that you had an exception related to that.

Please try https://dev.azure.com/seipharos/pharos/_build/results?buildId=210&view=artifacts&type=publishedArtifacts

I am optimistic that this will fix the problem for you!

**Comment by astrelsky** [comment]
@sei-eschwartz unfortunately failing to commit any sub transaction forces all concurrent transactions to be rolled back. This is the documented and intended behavior. https://ghidra.re/ghidra_docs/api/ghidra/framework/model/UndoableDomainObject.html#endTransaction(int,boolean)

**Comment by sei-eschwartz** [comment]
@astrelsky Thanks, good to know that is the intended behavior.

So what is the point of allowing sub-transactions if they can't be individually rolled back?

**Comment by astrelsky** [comment]
> @astrelsky Thanks, good to know that is the intended behavior.
> 
> So what is the point of allowing sub-transactions if they can't be individually rolled back?

I don't know. I opened an issue asking if it is possible to do so about a month ago but didn't receive a response. https://github.com/NationalSecurityAgency/ghidra/issues/1991

**Comment by 0xBEEEF** [comment]
> The pastebin logs expired. But fortunately I think there are only two sub-transactions and they are easy to get rid of. They are both related to adding types to the data type manager, and I do recall that you had an exception related to that.
> 
> Please try https://dev.azure.com/seipharos/pharos/_build/results?buildId=210&view=artifacts&type=publishedArtifacts
> 
> I am optimistic that this will fix the problem for you!

Many thanks for the feedback! I have downloaded and started it, it can only be hours.

**Comment by 0xBEEEF** [comment]
Great news! The last correction has now really brought it. The structures and classes have been created successfully. This means that one error has been corrected. Many thanks to both of you for the great and fast correction of the transaction error.

Now we can finally take care of the original one again, since the point was that a class with the same name exists in different namespaces. Here something gets lost during the import and the exceptions occur.

**Comment by astrelsky** [comment]
> Great news! The last correction has now really brought it. The structures and classes have been created successfully. This means that one error has been corrected. Many thanks to both of you for the great and fast correction of the transaction error.
> 
> Now we can finally take care of the original one again, since the point was that a class with the same name exists in different namespaces. Here something gets lost during the import and the exceptions occur.

Glad to hear the transaction issue has been resolved.

Now regarding the class with the same name in different namespaces. While there isn't really anything I can do to help with this one I would like to make it known that Ghidra may end up using the wrong Structure as the `this` parameter for some functions in this situation. Unfortunately there is absolutely nothing you can do about it and it has been reported with a suggested solution in [#1210](https://github.com/NationalSecurityAgency/ghidra/issues/1210)

**Comment by sei-eschwartz** [comment]
Glad the transaction issue is fixed.  I looked briefly and there are at least two cases where the class name matching (without the namespace) is helping:

* It helps us find the class name if it has already been moved into the OOAnalyzer namespace.
* It helps us find the class name if it is imported from a DLL.

I will try to handle these two situations while considering the namespace.

**Comment by sei-eschwartz** [comment]
I think I sorted this out.

https://dev.azure.com/seipharos/pharos/_build/results?buildId=213&view=artifacts&type=publishedArtifacts

**Comment by 0xBEEEF** [comment]
> I think I sorted this out.
> 
> https://dev.azure.com/seipharos/pharos/_build/results?buildId=213&view=artifacts&type=publishedArtifacts

Many thanks for the renewed correction. I'll test it tomorrow during the day. I'll get back to you with either good or bad news to see if that solves the problems completely.

**Comment by 0xBEEEF** [comment]
Not such good news, I'm afraid. The last changes concerning the namespaces did not work. The error still occurs. @sei-eschwartz , did you really build an example like the one I gave you above with the same structure? Something definitely seems not to work. The same exceptions still occur during import...

**Comment by sei-eschwartz** [comment]
No, I did not make an example.  I was proceeding under the assumption that the problem was caused by inexact class matching.

Can you look in the log for "Looking for a class symbol named XYZ" where XYZ is one of the TestClassT classes?

There should also be a nearby message saying `Symbol for class XYZ:` which reports the path it found.

The other possibility is that OOAnalyzer goofed up somehow.  Can you share the json for `TestClassT`?  Particularly the members.  Is there actually a cycle in the JSON?

**Comment by 0xBEEEF** [comment]
Unfortunately I didn't look into the log in detail, because I immediately saw the same exceptions again. To the file, yes, I had provided you with the excerpt above, but it has expired in the meantime. For this reason I have uploaded the excerpt [here](https://pastebin.com/imJdxmyS) again. This is everything around this data type and how it looks like. I think this should help you in any case with the analysis.

**Comment by sei-eschwartz** [comment]
I am posting the exerpt here so it does not expire 
[Test Json.txt](https://github.com/cmu-sei/pharos/files/5016853/Test.Json.txt)

**Comment by sei-eschwartz** [comment]
It does not look like there is a cycle from the exerpt.  I will check to see if perhaps we have another place that is doing "fuzzy matching".

**Comment by sei-eschwartz** [comment]
I was able to reproduce the problem by making a toy program.  Have not had a chance to investigate yet.

**Comment by sei-eschwartz** [comment]
Ok, I think I finally got it sorted out.  Although we set the correct category path at one point, we then moved each datatype to /OOAnalyzer instead of prepending /OOAnalyzer to the category path.

Try this: https://dev.azure.com/seipharos/pharos/_build/results?buildId=219&view=artifacts&type=publishedArtifacts

I also saw @0xBEEEF's comment in #109 that our data types could use some improvement in general and agree.

**Comment by 0xBEEEF** [comment]
> Ok, I think I finally got it sorted out. Although we set the correct category path at one point, we then moved each datatype to /OOAnalyzer instead of prepending /OOAnalyzer to the category path.
> 

Thank you very much. I'll try it first thing tomorrow. 👍 

By the way, I noticed something else regarding the VFTable search: Under certain circumstances it does not work in the (virtual) multiple inheritance. Apparently this is the case when there are multiple virtual base classes. Only the first VFTable is inserted and the others are seen as normal members. In Ghidra I see the link to the correct VFTable in the affected places, but it is not shown as such. What exactly is the reason for this I cannot say yet. I'm trying to recreate a case here, including the source code and open a new issue if I can. So this is just an info, there is nothing to do (yet). I might come back to this.

**Comment by sei-eschwartz** [comment]
Could you show the entries in the JSON for the vftables in question?  (It would be good to create a new issue for this.)

**Comment by 0xBEEEF** [comment]
Good news! All problems are fixed now. The originally mentioned error no longer occurs. From my point of view this issue can be closed when the changes have been merged to the master branch.

> Could you show the entries in the JSON for the vftables in question? (It would be good to create a new issue for this.)

A short note about the already announced new bug. I'll open a new issue for this one the next days. But first of all. The JSON file is not my fault, the bug occurs before that. So there is no bug in the actual Ghidra plugin.

**Comment by sei-eschwartz** [comment]
Hurray!
